Feature/medios contactos

HopeAero opened the pull request:
Medios de Contacto implementados

Esta rama contiene el feature/migration y hotfix/user-entity

LuisRivasW committed:
Metodos de contacto para que revise emmanuel+

HopeAero committed:
Merge branch 'epic/ultimate-v1-merging-branch' into mediosContactos

HopeAero committed:
Method Contact 100% implement

HopeAero committed:
Hotfix users crud

HopeAero committed:
Implement migration

HopeAero committed:
Migration complete

HopeAero committed:
Merge branch 'epic/ultimate-v1-merging-branch' into feature/mediosContactos

HopeAero committed:
Merge branch 'feature/migration' into feature/mediosContactos

HopeAero committed:
Merge branch 'hotfix/users-entity' into feature/mediosContactos

HopeAero committed:
fixes veloz

HopeAero committed:
Update file paths in package.json and
datasource.ts

HopeAero committed:
Merge branch 'feature/migration' into feature/mediosContactos

Hector1567XD commented on src/core/auth/dto/register.dto.ts:
```diff
@@ -13,6 +13,11 @@ export class RegisterDto {
+  @ApiProperty()
+  @IsString()
+  @MinLength(2)
+  document: string;
```
Ya hablamos de esto en otro PR

Hector1567XD commented on src/core/auth/service/auth.service.ts:
```diff
@@ -43,7 +43,7 @@ export class AuthService {
   ) {}
 
-  async register({ email, password, name, lastname }: RegisterDto) {
+  async register({ email, password, name, lastname, document }: RegisterDto) {
```
Hablamos de esto en otro PR, pero recuerda, debe ser unico, debe ser requerido y si ya existe debe decir "nao nao, ya existe"

Hector1567XD commented on src/core/profiles/dto/responses.dto.ts:
```diff
@@ -121,3 +122,48 @@ export class ResponsePaginationProfile {
+    type: [ExperienceData],
+  })
+  experience: Experience[];
+  @ApiProperty({
+    type: [SkillData],
+  })
+  skills: Skill[];
```
El tipo `SkillData` es distinto de `Skill` especificado mas abajo, ¿por que?
```suggestion
  @ApiProperty({
    type: [SkillData],
  })
  skills: Skill[];
```

HopeAero replied:
swagger se chispotea cuando una entidad tiene muchas relaciones y no puede hacer el ejem response asi que te recomienda hacer una clase para que la respuesta example funcione

Hector1567XD replied:
Tenemos que revisar esto juntos despues

Hector1567XD commented on src/core/profiles/dto/responses.dto.ts:
```diff
@@ -3,6 +3,7 @@ import { Profile } from '../entities/profile.entity';
 import { Skill } from 'src/core/skills/entities/skill.entity';
 import { Experience } from 'src/core/experience/entities/experience.entity';
 import { User } from 'src/core/users/entities/user.entity';
+import { ProfileI } from '../entities/profile.interface';
```
Se deberia extender o implementar algo de esta interface en este archivo? para que se importa?

Hector1567XD commented on src/core/profiles/dto/responses.dto.ts:
```diff
@@ -121,3 +122,48 @@ export class ResponsePaginationProfile {
+  value: string;
+}
+
+export class ResponseMethodContactDTO {
+  @ApiProperty({
+    type: ProfileData,
+  })
+  profile: ProfileData;
+  @ApiProperty({
+    type: [MethodContact],
+  })
+  contactMethods: MethodContact[];
+}
```
wtf o.0???? why???

Hector1567XD replied:
Habria que borrarlo no ?

Hector1567XD commented on src/core/profiles/dto/responses.dto.ts:
```diff
@@ -121,3 +122,48 @@ export class ResponsePaginationProfile {
+  skills: Skill[];
+}
+
+export class MethodContact {
+  @ApiProperty()
+  id: number;
+
+  @ApiProperty()
+  type: string;
+
+  @ApiProperty()
+  value: string;
+}
```
Por que volver a crear esto :S? por que no reutilizar el Entity `ContactMethod`?

Hector1567XD commented on src/core/profiles/messages.ts:
```diff
@@ -5,3 +5,5 @@ export const MY_PROFILE_SUCCESFULLY_UPDATED =
   'Skill eliminado de tu perfil exitosamente';
 export const ERROR_PROFILE_SKILL_NOT_FOUND = 'Perfil o skill no se encontraron';
+export const PROFILE_SUCCESFULLY_DELETE_METHOD_CONTACT =
+  'Contacto eliminado con exito';
```
```suggestion
  'Metodo de contacto eliminado con exito';
```

Hector1567XD commented on src/core/profiles/profiles.controller.ts:
```diff
@@ -187,4 +196,44 @@ export class ProfilesController {
+
+  @ApiTags('profile')
+  @Auth(UserRole.GRADUATE)
+  @ApiCreatedResponse({
+    description: 'Add contact method to my profile',
+    type: ResponseMethodContactDTO,
+  })
+  @Post('/my-profile/contact-methods')
+  async addContactMethod(
+    @ActiveUser() user: UserActiveInterface,
+    @Body() createContactMethodDto: CreateContactDto,
+  ) {
+    return this.profilesService.addContactMethod(user, createContactMethodDto);
+  }
+
+  @ApiTags('profile')
+  @Auth(UserRole.GRADUATE)
+  @Get('/my-profile/contact-methods')
+  async getContactMethods(@Param('id', ParseIntPipe) id: number) {
+    return this.profilesService.getContactMethods(id);
+  }
+
+  @ApiTags('profile')
+  @Auth(UserRole.GRADUATE)
+  @ApiOkResponse({
+    description: 'Delete contact method from my profile',
+    type: MessageDTO,
+  })
+  @Delete('/my-profile/contact-methods/:id')
+  async deleteContactMethod(
+    @Param('id', ParseIntPipe) id: number,
+    @ActiveUser() user: UserActiveInterface,
+    @Response() response: express.Response,
+  ) {
+    await this.profilesService.removeContactMethod(id, user);
+
+    return response.status(200).json({
+      message: PROFILE_SUCCESFULLY_DELETE_METHOD_CONTACT,
+    });
+  }
```
Definitivamente tienes que eliminar esto, te hiciste un CRUD cuando lo MENOS que buscaba era un CRUD
```suggestion
```

Hector1567XD commented on src/core/profiles/service/index.ts:
```diff
@@ -167,4 +169,61 @@ export default class ProfilesService {
+      where: { id: user.id },
+    });
+
+    if (!profile) {
+      throw new NotFoundException(PROFILE_NOT_FOUND);
+    }
+
+    const contactMethod = new ContactMethod();
+    contactMethod.id = profile.contactMethods.length + 1;
+    contactMethod.type = createContactMethodDto.type;
+    contactMethod.value = createContactMethodDto.value;
+
+    profile.contactMethods.push(contactMethod);
+
+    await this.profileRepository.save(profile);
+
+    return profile;
+  }
+
+  async removeContactMethod(
+    id: number,
+    user: UserActiveInterface,
+  ): Promise<void> {
+    const profile = await this.profileRepository.findOne({
+      where: { id: user.id },
+    });
+
+    if (!profile) {
+      throw new NotFoundException(PROFILE_NOT_FOUND);
+    }
+
+    const updatedContactMethodList = profile.contactMethods.filter(
+      (contactMethod) => contactMethod.id !== id,
+    );
+
+    profile.contactMethods = updatedContactMethodList;
+
+    await this.profileRepository.save(profile);
+    return;
+  }
```
Todo esto se va
```suggestion
```

Hector1567XD replied:
(Bueno, quiza no todo, pero lo vemos juntos despues de la daily)

Hector1567XD commented on src/core/users/dto/update-user.dto.ts:
```diff
@@ -14,8 +14,14 @@ import { Transform } from 'class-transformer';
 export class UpdateUserDto extends PartialType(CreateUserDto) {
```
Deberia haber una forma de reutilizar todo lo de CreateUserDto pero haciendo los campos opcionales no?... podrias googlear o chatgptear algo de eso, seguro nos ahorramos MUCHO pero MUCHISIMO codigo

Hector1567XD commented on src/core/users/entities/user.interface.ts:
```diff
@@ -1,5 +1,5 @@
 export interface UserI {
```
Le falta `document` no?

HopeAero replied:
nop

Hector1567XD commented on src/core/users/entities/user.interface.ts:
```diff
@@ -1,5 +1,5 @@
-  email?: string;
+  name: string;
+  lastname: string;
+  email: string;
 }
```
```suggestion
  email: string;
  document: string;
}
```
?

HopeAero committed:
Document change name

HopeAero committed:
Refactor profile and user DTOs and entities

This commit refactors the profile and user DTOs and entities. It adds a unique constraint to the documentNumber field in both the CreateUserDto and UpdateUserDto classes. Additionally, it adds a unique constraint to the documentNumber column in the User entity.

HopeAero committed:
Fixes post reu FI

Yisustxz reviewed (approved):
todo piola

Nannajpg reviewed (commented):
/api/v1/profiles/my-profile/contact-methods esta rarongo

HopeAero committed:
Remove unused getContactMethods method

HopeAero commented:
> /api/v1/profiles/my-profile/contact-methods esta rarongo

correginho

Hector1567XD commented on src/core/profiles/dto/responses.dto.ts:
```diff
@@ -121,3 +121,45 @@ export class ResponsePaginationProfile {
+  email: string;
+}
+
+export class ResponseMethodContactDTO {
```
Acabo de ver, esta es la respuesta cuando modificas los metodos de contacto del perfil, me parace genial devuelvas el Perfil con sus contact-methods actualizados en ese caso pero.... deberias reutilizar el DTO que ya hay para la respuesta de Perfil y llenarle los datos faltantes si hacen falta para cumplirlos

Hector1567XD replied:
Podrias reutilizar el servicio de profile para enviar respuesta

luizzzito reviewed (approved):
Los endpoints funcionan de forma correcta, y no creo que deban haber cambios en las respuestas

HopeAero committed:
Add contact method to profile
